Add RocksDB stats

cuongdo opened the pull request:
This adds a variety of store-level RocksDB stats with the goal of better
understanding the memory and performance characteristics of RocksDB.
For now, these stats are recorded to time series and are viewable
through `/_status/metrics/NODEID`.

Also fixes a problem with a call to `MarshalJSON` while servicing
`/_status/metrics/NODEID`.

Resolves #1179.

cc @petermattis for RocksDB aspects, @mrtracy for stats backend

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/5099)

<!-- Reviewable:end -->

Commit:
Add RocksDB stats

This adds a variety of store-level RocksDB stats with the goal of better
understanding the memory and performance characteristics of RocksDB.
For now, these stats are recorded to time series and are viewable
through `/_status/metrics/NODEID`.

Also fixes a problem with a call to `MarshalJSON` while servicing
`/_status/metrics/NODEID`.

Resolves #1179.

mrtracy commented:
From what I can read on my phone, the metrics portion in the Store is LGTM.  I'm not sure what other test could be written that would provide significant additional value -  perhaps a test that the server is actually recording values for all of the stores gauges?  That is super high level (above the level of this change) , but it is missing. Probably just log an issue for that.

cuongdo commented:
Review status: 4 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[server/status/recorder.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX5DHZql0sAREko28j:-KCYeQ3XenGN_84bPlDC:740735385) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/server/status/recorder.go#L172)):*
The test is `TestMetricsEndpoint` in `server/status_test.go`. Yeah, I'll separate out in a separate PR next time.

Changed the conversion to `strconv.Itoa`.

---

*[storage/engine/engine.go, line 170 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX7KOTsqWS6WsxQW2o:-KCYkfUbrUTvNZ7lAwhR:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/engine/engine.go#L170)):*
Done.

---

*[storage/engine/rocksdb.go, line 454 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX60Ln9avll3tOj5RZ:-KCYkwXF6gB0x9lSt0dU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/engine/rocksdb.go#L454)):*
Done.

---

*[storage/engine/rocksdb/db.h, line 180 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX7Td8SdI75kTQa8Jh:-KCYlkA95OaRM3y1U3LU:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/engine/rocksdb/db.h#L180)):*
Done.

---

*[storage/engine/rocksdb/eventlistener.cc, line 1 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX6zFT4eL9dPC2V4ZX:-KCYlptuRCbcaHZPF5BN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/engine/rocksdb/eventlistener.cc#L1)):*
Done.

---

*[storage/engine/rocksdb/eventlistener.cc, line 20 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX7Au8-eXqKHAhe6dc:-KCYlvYfosGCf7hOON7M:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/engine/rocksdb/eventlistener.cc#L20)):*
Done.

---

*[storage/store_test.go, line 1729 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX5huNFE9ord1vi6oX:-KCYkJZWvSyfJbvcsHfK:-1235581457) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/storage/store_test.go#L1729)):*
I've added another test to `StoreMetrics`, which is more comprehensive. Since that new test is a superset of this test, I've removed this one.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 7 files at r3, 2 of 2 files at r4.
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/status/recorder.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX5DHZql0sAREko28j:-KCZ06gIDwzQFqc9xLy6:-1024676289) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/server/status/recorder.go#L172)):*
What is `TestMetricsEndpoint` testing? It doesn't appear to care if string keys are used. What am I missing?

Also, I didn't mean to suggest a separate PR was needed - a separate commit in this same PR would have been fine too.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[server/status/recorder.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX5DHZql0sAREko28j:-KCZ1yJODJrbX-EwbGkA:1873315407) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/server/status/recorder.go#L172)):*
It's testing that `/_status/metrics/NODEID` can be retrieved with no HTTP errors. Previously, `statusServer.handleMetrics()` was failing to marshal this map into JSON because of the non-string key. That resulted in an HTTP 500 Internal Server Error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[server/status/recorder.go, line 172 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCX5DHZql0sAREko28j:-KCZ3CzfauF8FSBAevBe:-509259079) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c881096f69069d18c5217e4270febfcb923d613/server/status/recorder.go#L172)):*
Ah, I didn't realize the JSON marshalling would actually fail. Thanks.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 12 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-:-KCa3U1BVDGTobB6ukSP:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 8 of 12 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb.go, line 559 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCaChR-XMODthPLumCN:-KCaChR-XMODthPLumCO:-1172714100) ([raw file](https://github.com/cockroachdb/cockroach/blob/70931f9c1b2e766f8fefc4c59139dae8d389ceb0/storage/engine/rocksdb.go#L559)):*
other places just panic.

---

*[storage/engine/rocksdb/eventlistener.cc, line 25 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCaCt1V_6CJa5aoDRfI:-KCaCt1WkjS-QuX1g9Ho:-1142898872) ([raw file](https://github.com/cockroachdb/cockroach/blob/70931f9c1b2e766f8fefc4c59139dae8d389ceb0/storage/engine/rocksdb/eventlistener.cc#L25)):*
no need for synchronization here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 4 of 12 files at r1, 1 of 1 files at r2, 6 of 7 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[storage/engine/rocksdb/eventlistener.cc, line 25 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099#-KCaCt1V_6CJa5aoDRfI:-KCaDT03lEDXcwnOPVsd:-1934729791) ([raw file](https://github.com/cockroachdb/cockroach/blob/70931f9c1b2e766f8fefc4c59139dae8d389ceb0/storage/engine/rocksdb/eventlistener.cc#L25)):*
C++ magic: these are `std::atomic` variables. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/5099)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Nice to see these added!
